Requiring Dasher Admin Status as a Decorator for Using the Management Server

xwsxethan opened the pull request:
In order to implement this, we require some kind of oauth so that we can query the directory API for admin status. In the future, we could lessen the oauth grant to require only user access (since that is all that is needed to verify an admin) on most pages. The exception would be the add_user page which would still require group and member access (read only) in order to allow users to be imported via group. For now, all access is required for everything, which doesn't make things easier for users, but it does make us more secure until we implement different oauth decorators for different access levels.

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/uproxy/ufo-management-server/45)
<!-- Reviewable:end -->

blueandgold commented on proxy_server.py:
```diff
@@ -59,12 +61,16 @@ class AddProxyServerHandler(webapp2.RequestHandler):
 
   @admin.RequireAdmin
+  @OAUTH_DECORATOR.oauth_required
+  @dasher_admin.DasherAdminAuthRequired
```
I think we should combine the 2 different admin decorators into one.  For example, the @admin.RequireAdmin can be made to look like below, and we can add a IsDomainAdmin(user) method in google_directory_service.  This will make the code much cleaner.

Also, should change "dasher" to "domain" so that it's easier for external users to understand.

    if not user:
      logging.error('user is not logged in')
      self.abort(403)
    if not users.is_current_user_admin():
      logging.error('user is not an admin')
      self.abort(403)
   if not directory_service.IsDomainAdmin(user):
      logging.error('user is not an domain admin')
      self.abort(403)

blueandgold commented on dasher_admin.py:
```diff
@@ -0,0 +1,46 @@
+      logging.error('Exception when asking dasher for this user.')
+      self.abort(403)
+
+    if not directory_user or directory_user is None:
```
Would this condition check be necessary, after a user has logged in?

xwsxethan commented:
Reviewed 5 of 5 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[dasher_admin.py, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZt-r1-37:-K577XWoSjWYNkSwBeCO:1324579868) ([raw file](https://github.com/uproxy/ufo-management-server/blob/7c84babe2810c63925fa796d93ef1f0e32902fd4/dasher_admin.py#L37)):*
It is still technically possible for the directory service call to return empty. It would be a weird scenario though, so it probably will never be hit, but I have it in here just to be safe.

---

*[proxy_server.py, line 65 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZp-r1-65:-K577uK68oMIRWE4cEtT:-1389174557) ([raw file](https://github.com/uproxy/ufo-management-server/blob/7c84babe2810c63925fa796d93ef1f0e32902fd4/proxy_server.py#L65)):*
I like what you're thinking, but I'm not sure if combining it with admin is the right thing to do. We still want the admin decorator for the setup flow so that we can at least guarantee something about the user on initial configuration when we can't do any OAuth.

I'm thinking we may eventually want to do two subclasses of OAuth decorators. One that requests just user scope (in the base class) and then checks for admin (in the subclass) as well as another one that requests all scopes and then checks for admin.

Would that be cleaner in your opinion? It definitely looks a bit messy with all the decorators currently so we need to do something.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45)*
<!-- Sent from Reviewable.io -->

xwsxethan commented:
Reviewed 10 of 10 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[proxy_server.py, line 65 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZp-r1-65:-K5BWoq20JtXyoJvys_x:-1115752312) ([raw file](https://github.com/uproxy/ufo-management-server/blob/7c84babe2810c63925fa796d93ef1f0e32902fd4/proxy_server.py#L65)):*
I couldn't get the inheritance to work properly with a decorator, so oauth had to stay. However, that's probably not the worst thing in the world since we may not want to mess with oauth ourselves anyway. So the oauth and xsrf decorators still exist as before. I moved the oauth decorators (now there are separate ones for just user scope and all scopes) into the admin file just so it is all in one place. I implemented the double check of domain and application admin as necessary though.

---

*[dasher_admin.py, line 37 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZt-r1-37:-K5BXIzyxBcRCxNDBEeH:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server/blob/7c84babe2810c63925fa796d93ef1f0e32902fd4/dasher_admin.py#L37)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45)*
<!-- Sent from Reviewable.io -->

xwsxethan commented:
Landscape IO thinks I introduced poor code quality, but actually I just move a TODO from one file to another, and it hasn't realized that the TODO is the same as before, so there should be no overall change. You can verify yourself if necessary.

blueandgold commented on admin.py:
```diff
@@ -25,3 +37,55 @@ def decorate(self, *args, **kwargs):
+    client_secret=OAuth.GetOrInsertDefault().client_secret,
+    scope=SCOPES)
+
+OAUTH_USER_SCOPE_DECORATOR = OAuth2Decorator(
```
Is there a reason to break out the user scope into a separate decorator?  I can see this can help to tighten access.  But this would prompt the user a couple of times to approve the oauth access?  I read somewhere (I can't find it now) that the better practice is for the app to prompt the user once to authorize all the scopes that the app needs.  And maintaining separate decorators seem to be making the code more complicated.  Would it be possible to just maintain one oauth decorator?

blueandgold commented on proxy_server.py:
```diff
@@ -58,12 +56,14 @@ class AddProxyServerHandler(webapp2.RequestHandler):
 
-  @admin.RequireAdmin
+  @admin.OAUTH_USER_SCOPE_DECORATOR.oauth_required
+  @admin.RequireAppAndDomainAdmin(admin.OAUTH_USER_SCOPE_DECORATOR)
```
I would hope to see something simple like one decorator like below, which would tell me exactly what check is required before processing this method. 
@admin.RequireAppAndDomainAdmin

Instead, the use of the two decorators here is really not intuitive.  I have to go back and see where they are defined to understand what is happening.

blueandgold commented on proxy_server.py:
```diff
@@ -165,4 +169,8 @@ def get(self):
     ('/proxyserver/list', ListProxyServersHandler),
 
     ('/cron/proxyserver/distributekey', DistributeKeyHandler),
+    (admin.OAUTH_ALL_SCOPES_DECORATOR.callback_path,
```
Seems odd to have all these path routing.

blueandgold commented on admin.py:
```diff
@@ -25,3 +37,55 @@ def decorate(self, *args, **kwargs):
+        kwargs: Parameters passed on to the specified function if successful.
+      """
+      user = users.get_current_user()
+      if not user or user is None:
```
Can this chuck of code be pulled out into something like below, so that it can be reused by both of the require admin methods?

def AbortIfUserIsNotLoggedIn(user):
     if not user:
       logging.error('user is not logged in')	
       self.abort(403)

blueandgold commented on admin.py:
```diff
@@ -25,3 +37,55 @@ def decorate(self, *args, **kwargs):
+      directory_user = None
+      try:
+        directory_service = GoogleDirectoryService(oauth_decorator)
+        directory_user = directory_service.GetUser(identifier)
```
It would seem nicer to have the move this logic into the directory service, such as:
directory_service.IsAdminUser(user)

xwsxethan commented:
Review status: 4 of 10 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[admin.py, line 53 \[r2\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K5BVM0XEcMA_XYyOrkG-r2-53:-K5BwxaC5QKlsHt9XtqY:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server/blob/9ec59d249a881513bc8da1218170f8e684cc8556/admin.py#L53)):*
Done.

---

*[admin.py, line 65 \[r2\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K5BVM0XEcMA_XYyOrkG-r2-65:-K5Bx-EIrFAK_u619SSt:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server/blob/9ec59d249a881513bc8da1218170f8e684cc8556/admin.py#L65)):*
Done.

---

*[admin.py, line 87 \[r2\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K5BVM0XEcMA_XYyOrkG-r2-87:-K5Bx2-nkCkfjEttk-PT:-665491744) ([raw file](https://github.com/uproxy/ufo-management-server/blob/9ec59d249a881513bc8da1218170f8e684cc8556/admin.py#L87)):*
The reason was to not request additional access than would be required at a given time. For instance, if you never run the import flow, then you shouldn't need to grant group or member access since that is the only place they are used. However, I can see your point about it being a worse user experience to have to perform multiple grants in different places.

In general it seems like this model of granting permissions isn't really that great since we have to ask for all these permissions that aren't really necessary for 90% of the app. I've changed it as you requested though.

---

*[proxy_server.py, line 172 \[r2\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZp-r2-172:-K5BxniEUvm1MYWmyPAj:-499146009) ([raw file](https://github.com/uproxy/ufo-management-server/blob/9ec59d249a881513bc8da1218170f8e684cc8556/proxy_server.py#L172)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45)*
<!-- Sent from Reviewable.io -->

blueandgold commented on admin.py:
```diff
@@ -15,13 +65,25 @@ def decorate(self, *args, **kwargs):
+    user_is_admin = False
+    try:
+      directory_service = GoogleDirectoryService(OAUTH_DECORATOR)
+      user_is_admin = directory_service.IsAdminUser(identifier)
```
nit: "user_is_admin" should rather be "is_admin_user" to be more consistent with boolean naming

blueandgold commented on proxy_server.py:
```diff
@@ -59,12 +57,14 @@ class AddProxyServerHandler(webapp2.RequestHandler):
 
-  @admin.RequireAdmin
+  @admin.OAUTH_DECORATOR.oauth_required
+  @admin.RequireAppAndDomainAdmin
```
Cool, this looks much cleaner than before!  

Just one small question: does the ordering of these decorators matter? If the decorators are called from the bottom up, and RequireAppAndDomainAdmin uses OAUTH_DECORATOR, then it's okay to call @admin.OAUTH_DECORATOR.oauth_required afterwards?

blueandgold commented:
:+1:   Cool, this looks much cleaner than before!  Just a nit and a small question.

xwsxethan commented:
Reviewed 6 of 6 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[proxy_server.py, line 65 \[r1\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZp-r1-65:-K5IfVOv3PQUnukvxV06:-1946685863) ([raw file](https://github.com/uproxy/ufo-management-server/blob/7c84babe2810c63925fa796d93ef1f0e32902fd4/proxy_server.py#L65)):*
I'm gonna take another stab at this at some point actually given that I found out how to do it. See the other comment.

---

*[proxy_server.py, line 61 \[r3\]](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45#-K56ri6Up1SOQw0hnZZp-r3-61:-K5IfhvVBEQ0KM9s-o3c:1585979857) ([raw file](https://github.com/uproxy/ufo-management-server/blob/728fab1f8131d96bd2a1d91001417bdba7d3eced/proxy_server.py#L61)):*
So I went looking for an answer to this, and boy did I hit the mother load of all decorator answers: https://stackoverflow.com/questions/739654/how-can-i-make-a-chain-of-function-decorators-in-python/1594484#1594484

They are executed top down and order does matter for the short version so doing it the way I have is correct. Given all this nifty help though, I'll take another shot at doing the decorator inheritance. That post should give me everything I need to make it work correctly. Gonna take a while though, so I'm not sure if I'll submit this and follow up with another PR, or just wait and submit it all at once. When I merge and resolve for all the other recent changes I'll let you know what I decide.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/uproxy/ufo-management-server/45)*
<!-- Sent from Reviewable.io -->
